### PR TITLE
fix(assistant-stream): skip emitted array entries while streaming

### LINKED ISSUE #7316: assistant-stream: streamed array readers revisit every emitted item
## Problem

`args.forEach("items")` starts scanning at array index zero on each update, including all items it already emitted. With one completed item per chunk, the number of membership checks grows quadratically:

| Streamed items | Membership checks |
| --- | ---: |
| 100 | 5,050 |
| 1,000 | 500,500 |

The output is correct and each item is emitted once. The issue is repeated work while consuming a growing array, separate from the cost of parsing the JSON.

## Reproduction

Reproduced against main at `8d39b458ec62b7e3cc4adb399fb960c71505cb32` (assistant-stream 0.3.42), using Node 24.21.0 and pnpm 12.3.4.

In a fresh clone of this repository, check out that commit and run `pnpm install --frozen-lockfile`. Save the following as `packages/assistant-stream/src/core/tool/foreach-cost.repro.test.ts`:

```ts
import { expect, it, vi } from "vitest";
import { ToolCallReaderImpl } from "./ToolCallReader";

it.each([100, 1000])(
  "repro: counts membership checks for %i streamed array entries",
  async (count) => {
    const reader = new ToolCallReaderImpl<{ items: number[] }, string>();
    const stream = reader.args.forEach("items");
    const original = Set.prototype.has;
    let membershipChecks = 0;
    const spy = vi.spyOn(Set.prototype, "has").mockImplementation(function (
      this: Set<unknown>,
      value: unknown,
    ) {
      if (typeof value === "number") membershipChecks++;
      return original.call(this, value);
    });
    try {
      await reader.appendArgsTextDelta('{"items":[');
      for (let i = 0; i < count; i++) {
        await reader.appendArgsTextDelta(
          String(i) + (i === count - 1 ? "]}" : ","),
        );
      }
      await reader.finishArgsText();
    } finally {
      spy.mockRestore();
    }
    const values = [];
    for await (const value of stream) values.push(value);
    expect(values).toEqual(Array.from({ length: count }, (_, i) => i));
    console.info(JSON.stringify({ count, membershipChecks }));
    expect(membershipChecks).toBe((count * (count + 1)) / 2);
  },
);
```

Run from the repository root:

```sh
pnpm -C packages/assistant-stream exec vitest run src/core/tool/foreach-cost.repro.test.ts
```

Both measurement tests pass and assert the counts above, while also asserting that every array item is returned in order. These are operation counts, not a wall-time benchmark or a claimed speedup for the whole reader.

## Root cause

[ForEachHandle.update](https://github.com/assistant-ui/assistant-ui/blob/8d39b458ec62b7e3cc4adb399fb960c71505cb32/packages/assistant-stream/src/core/tool/ToolCallReader.ts#L240-L254) loops over the entire array and checks `processedIndexes.has(i)` on every update. The set prevents duplicate emission but does not prevent rescanning the completed prefix.

## Proposed fix

Track the next unprocessed index, or otherwise revisit only the unfinished suffix, while preserving the existing completion rules. Add a deterministic work-count contract alongside tests for nested partial items, late subscription, cancellation, empty arrays and exactly-once ordered emission.

No API change appears necessary, but the partial-JSON completion rules need validation before using a monotonic cursor. #7193 changes the parser but leaves this loop in place, so the fixes should be coordinated.

<a href="https://rupic.dev/s/s2.JzjTOduMTvORRsZN3BqxIb5W7Z22MgTz_EcPHUOXJo7gNVtoan26gkLaMP0?ref=github" target="_blank" rel="noopener noreferrer"><picture><source media="(prefers-color-scheme: dark)" srcset="https://rupic.dev/badges/track-dark-v1.svg">Track in Rupic</picture></a>

### PR DESCRIPTION
**Pull request by Kinfe123**
## Problem

`args.forEach()` rescans already-emitted array entries on every streamed update. Output is correct, but consuming an array that grows by one item per chunk causes quadratic loop work. Reproduced on main at `8d39b458e` with `assistant-stream` 0.3.42, as reported in #7316.

Fixes #7316.

## Root cause

The processed-index set prevents duplicate emission but the loop still starts at zero each time.

## Change

Replace the set with the next unprocessed index. Stop at the unfinished trailing item and resume there on the next update. A streamed JSON array has a completed prefix and, at most, an incomplete trailing element.

A deterministic contract counts actual array-bound reads through a proxy around the real parser output. For 100 items the count drops from 5,151 to 201; for 1,000 items it drops from 501,501 to 2,001. These are operation counts, not a claimed whole-reader wall-time speedup.

Tests also cover nested partial items, late subscription, cancellation, empty and missing arrays, truncated input and exactly-once ordered output. The parser itself is unchanged, keeping this separate from #7193.

## Verification

Both work-count tests fail against the old loop and pass with the cursor; the 6 output/lifecycle cases pass on both versions.

- `pnpm -C packages/assistant-stream exec vitest run src/core/tool/ToolCallReader.test.ts src/core/tool/ToolCallReader.forEach.test.ts`: 21 tests pass.
- `pnpm -C packages/assistant-stream exec vitest run`: 567 tests pass; 22 Redis integration tests skip without a live Redis server.
- `pnpm exec turbo run build --filter=assistant-stream... --output-logs=errors-only`: passes.
- Oxlint on changed files, oxfmt, `pnpm changesets:check` and `git diff --check`: pass.
- The repository size checker passes for the built entries, using the existing checker from a fully installed checkout against this branch's dist and budgets. No budget update needed.

The repository's `test:types` command, scoped to `assistant-stream`, passes.

Local verification used Node 24.21.0 and a frozen, package-filtered install after unrelated large workspace downloads timed out. Commands used `--config.verifyDepsBeforeRun=false` to keep the verification scope fixed. The formatter reused the cached Tailwind dependency; no docs files were changed.

## Public surface

`assistant-stream`: implementation-only fix and a patch changeset. No exports, signatures, templates or documentation claims change. No hand-edited API-reference output.

<a href="https://cubic.dev/pr/assistant-ui/assistant-ui/pull/7327?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

<a href="https://rupic.dev/s/s2.WhvAxGgsloF5aUCY9LDzVMrPBW-7FFVZzJV6i2KANZoommLilSM_tlhO-R8?ref=github" target="_blank" rel="noopener noreferrer"><picture><source media="(prefers-color-scheme: dark)" srcset="https://rupic.dev/badges/track-dark-v1.svg">Track in Rupic</picture></a>

### CODE REVIEW

File: packages/assistant-stream/src/core/tool/ToolCallReader.forEach.test.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+  it("waits for nested partial elements and emits them once in order", async () => {
+    const reader = new ToolCallReaderImpl<
+      { items: { label: string; tags: string[] }[] },
+      string
+    >();
+    const values = reader.args.forEach("items").getReader();
+    const first = values.read();
+    let settled = false;
+    void first.then(() => {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/assistant-stream/src/core/tool/ToolCallReader.forEach.test.ts]: Added the count breakdown in 671b6da9f: one bound check for the initial empty array, then two per completed item. The two reader suites pass together with 21 tests.

File: packages/assistant-stream/src/core/tool/ToolCallReader.forEach.test.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+      const values = reader.args.forEach("items");
+      await reader.appendArgsTextDelta('{"items":[');
+      for (let index = 0; index < count; index++) {
+        await reader.appendArgsTextDelta(
+          String(index) + (index === count - 1 ? "]}" : ","),
+        );
+      }
+      await reader.finishArgsText();
+
+      expect(arrayLengthReads).toHaveBeenCalledTimes(2 * count + 1);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/assistant-stream/src/core/tool/ToolCallReader.forEach.test.ts:57]: I checked this against the installed Vitest 5.0.0 defaults. clearMocks is true by default, and this package does not override it. Both parameterized cases pass together, as does the full suite (567 passed, 22 Redis tests skipped). No extra reset is needed here.

**Comment by Kinfe123** [comment]
Moved the six behavioral cases into the normal reader suite in 671b6da9f, so only the two count contracts use the parser proxy. Also documented the counter breakdown and the complete-prefix invariant. Focused tests, the full package suite, scoped type check, lint and formatting pass.

On the scope question, #7316 explicitly tracks prefix rescans separately from JSON parsing, so the closing link still fits. Parsing remains separate work in #7193; this PR does not claim a whole-reader timing speedup.
